operator/src/main/java/oracle/kubernetes/operator/DomainStatusUpdater.java

```diff
@@ -787,71 +560,24 @@ private FailedStep(Throwable throwable) {
     public NextAction apply(Packet packet) {
       LOGGER.entering();
 
-      boolean madeChange = false;
-
-      DateTime now = DateTime.now();
-      DomainPresenceInfo info = packet.getSPI(DomainPresenceInfo.class);
+      DomainConditionStepContext context = new DomainConditionStepContext(packet);
+      final DomainStatus status = context.getStatus().clearModified();
 
-      Domain dom = info.getDomain();
-      DomainStatus status = dom.getStatus();
-      if (status == null) {
-        status = new DomainStatus();
-        status.setStartTime(now);
-        dom.setStatus(status);
-        madeChange = true;
-      }
-
-      List<DomainCondition> conditions = status.getConditions();
-      if (conditions == null) {
-        conditions = new ArrayList<>();
-        status.setConditions(conditions);
-      }
-
-      ListIterator<DomainCondition> it = conditions.listIterator();
-      boolean foundFailed = false;
-      while (it.hasNext()) {
-        DomainCondition dc = it.next();
-        switch (dc.getType()) {
-          case FAILED_TYPE:
-            foundFailed = true;
-            if (!TRUE.equals(dc.getStatus())) {
-              dc.setStatus(TRUE);
-              dc.setReason("Exception");
-              dc.setMessage(throwable.getMessage());
-              dc.setLastTransitionTime(now);
-              madeChange = true;
-            }
-            break;
-          case PROGRESSING_TYPE:
-            if (!FALSE.equals(dc.getStatus())) {
-              dc.setStatus(FALSE);
-              dc.setLastTransitionTime(now);
-              madeChange = true;
-            }
-            break;
-          case AVAILABLE_TYPE:
-            break;
-          default:
-            it.remove();
-            madeChange = true;
-        }
-      }
-      if (!foundFailed) {
-        DomainCondition dc = new DomainCondition();
-        dc.setType(FAILED_TYPE);
-        dc.setStatus(TRUE);
-        dc.setReason("Exception");
-        dc.setMessage(throwable.getMessage());
-        dc.setLastTransitionTime(now);
-        conditions.add(dc);
-        madeChange = true;
+      status.addCondition(
+          new DomainCondition(Failed)
+              .withStatus(TRUE)
+              .withReason("Exception")
+              .withMessage(throwable.getMessage()));
+      if (status.hasConditionWith(c -> c.hasType(Progressing))) {
+        status.addCondition(new DomainCondition(Progressing).withStatus(FALSE));
       }
 
-      LOGGER.info(MessageKeys.DOMAIN_STATUS, dom.getDomainUID(), status);
+      LOGGER.info(MessageKeys.DOMAIN_STATUS, context.getDomain().getDomainUID(), status);
       LOGGER.exiting();
 
-      return madeChange == true
-          ? doDomainUpdate(dom, info, packet, FailedStep.this, getNext())
+      return status.isModified()
+          ? doDomainUpdate(
+              context.getDomain(), context.getInfo(), packet, FailedStep.this, getNext())
```

Comment:
seems to be a lot of changes in this file - is this just a bug fix?

Reply:
Some came from CheckStyle and the rest from Russ' refactoring.